### PR TITLE
[KEYCLOAK-16723] Upgrade to Wildfly 22

### PR DESCRIPTION
**Pull request by iankko**
[KEYCLOAK-16723] Upgrade to Wildfly 22
    
    Base fixes:
    * [KEYCLOAK-16724] Upgrade Keycloak to Wildfly 22.0.0.Beta1 / Wildfly Core 14.0.0.Beta4
    * [KEYCLOAK-16822] Upgrade Keycloak to Wildfly 22.0.0.Final / Wildfly Core 14.0.0.Final
    * [KEYCLOAK-17158] Upgrade Keycloak to Wildfly 22.0.1.Final / Wildfly Core 14.0.1.Final
    
    Other (related) fixes:
    * [KEYCLOAK-16174] Upgrade (RH-SSO adapters) to EAP CD 21
    * [KEYCLOAK-16202] Don't upgrade versions of httpclient and httpcore in the Fuse adapter
                       as part of the Wildfly upgrade script run
    * [KEYCLOAK-16737] Keycloak core depends on org.bouncycastle:bcprov-jdk15on:1.65, which
                       suffers from CVE-2020-28052
    * [KEYCLOAK-16907] ConfigMigrationTest fails after upgrade to Wildfly 22.0.0.Final
    * [KEYCLOAK-17156] org.keycloak.test.config.migration.ConfigMigrationTest fails with
                       'illegal reflective access to method com.sun.net.ssl.internal.ssl.Provider.isFIPS()'
    
    Signed-off-by: Jan Lieskovsky <jlieskov@redhat.com>

_Note:_ Needs to be merged together with corresponding:
* [Documentation](https://github.com/keycloak/keycloak-documentation/pull/1123) and
* [Website](https://github.com/keycloak/keycloak-web/pull/163)

changes.

@abstractj @hmlnarik @mhajas @martin-kanis @mposolda @pedroigor @pskopek @slaskawi @stianst @vmuzikar @vramik PTAL once got a chance

Thanks, Jan

### CODE REVIEW

**Comment by iankko** [comment]
This is now ready for review. The GHA tests above failed due to connection reset trying to download some Wildfly artifact (I doesn't seem to have the rights to rerun the jobs). Here's the correspond GHA run on my repo with the same commit SHA:
* https://github.com/iankko/keycloak/actions/runs/599189141

There are two test failures in that link:
* [MetricsRestServicesTest failure](https://github.com/iankko/keycloak/runs/1978454693?check_suite_focus=true) This is tracked under [KEYCLOAK-17231](https://issues.redhat.com/browse/KEYCLOAK-17231), scheduled to be fixed within ```Upgrade to Wildfly 23``` work,
* [ActionTokenCrossDCTest failure](https://github.com/iankko/keycloak/runs/1978454786?check_suite_focus=true) Will file a JIRA for this one yet.

The rest of the tests are passing.

**Comment by iankko** [comment]
See KEYCLOAK-16723 JIRA for links to corresponding adapters & server test pipeline runs.

File: distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-domain-clustered.cli [hunk 1]
```diff
@@ -72,7 +72,7 @@ set persistenceProvider=jpa
 if (outcome == failed) of /profile=$clusteredProfile/subsystem=infinispan/cache-container=keycloak/distributed-cache=authorization/:read-resource
     # In migration from 3.0.0 to 3.2.0 there is authorization distributed-cache replaced with local-cache
     try
-        echo
```
Comment (slaskawi) [inline distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-domain-clustered.cli:75]: Just double checking - this is intentional? I'm not sure an empty `echo` (without a space) is correct. Maybe it is - just double checking.

**Review by slaskawi — DISMISSED** [review]

File: distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-domain-clustered.cli (hunk 1 above)
Comment (iankko) [reply distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-domain-clustered.cli:75]: Yes, it is intentional to remove whitespace noise, present in migration to previous versions, where necessary / appropriate.

For the format itself, the new entries follow the previous format (one empty `echo` / `echo` without args for better readability between entries).

**Comment by pedroigor** [comment]
@iankko Do you know if the fix for the ISPN subsystem performance is in 22 ?

**Comment by pedroigor** [comment]
@iankko Re-running jobs.

**Comment by iankko** [comment]
@pedroigor 
> @iankko Do you know if the performance fix for the ISPN subsystem is in 22 ?

Hey Pedro, 

  sorry, which specific ISPN JIRA do you mean? (this PR would upgrade Infinispan from 11.0.3 to 11.0.8 AFAICT, so better to have specific JIRA ID, than to check all those releases)

Thanks,
Jan

**Comment by pedroigor** [comment]
@iankko Sorry. I already checked and the issue should be fixed. Here is https://issues.redhat.com/browse/WFLY-14356.

IMO, it would be nice to make sure it actually fixes the problem.

**Comment by martin-kanis** [comment]
@iankko thanks for the PR. It looks good to me.

@hmlnarik What do you think regarding the performance testing on our side? Do we want to re-run the performance testsuite to double check the problem is gone?

**Review by vramik — COMMENTED** [review]
Thank you @iankko for the PR, changes look good to me, I'd like to make sure migration tests passed with this PR. Do we have links to pipeline runs?

**Comment by iankko** [comment]
> Thank you @iankko for the PR, changes look good to me, I'd like to make sure migration tests passed with this PR. Do we have links to pipeline runs?

@vramik Hey Vlasto, thank you for looking into this! They [are added as a comment to KEYCLOAK-16723 JIRA](https://issues.redhat.com/browse/KEYCLOAK-16723?focusedCommentId=15848122&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15848122) (full links due to the recent Jenkins server migration).

Let me know, if they contain what you have been searching for (if not yet, will assure the results of those tests are in the next run if necessary).

Thanks, Jan

**Review by vramik — DISMISSED** [review]
@iankko thank you, there are some test failures in MigrationTest, but these are not related to this PR, so from my POV (migration) it's good to go.

**Comment by pedroigor** [comment]
@iankko We need the last commit from this branch [1] to properly update to newer ISPN versions.

It is basically https://issues.redhat.com/browse/WFLY-14356, but applied to our revision caches and when setting up the embedded caches.

It should be enough to fix the unnecessary byte[] conversion issue that we found when running performance tests some time ago.

[1] https://github.com/pedroigor/keycloak/tree/KEYCLOAK-16724

**Comment by iankko** [comment]
> @iankko We need the last commit from this branch [1] to properly update to newer ISPN versions.

@pedroigor Thanks!, looking. 
@stianst Please refrain from further processing this one for couple of mins yet, till I fixed the issue above. Thanks!

> 
> It is basically https://issues.redhat.com/browse/WFLY-14356, but applied to our revision caches and when setting up the embedded caches.
> 
> It should be enough to fix the unnecessary byte[] conversion issue that we found when running performance tests some time ago.
> 
> [1] https://github.com/pedroigor/keycloak/tree/KEYCLOAK-16724

**Comment by iankko** [comment]
> > @iankko We need the last commit from this branch [1] to properly update to newer ISPN versions.
> 
> @pedroigor Thanks!, looking.

@pedroigor  Done. Should you need anything else WRT to that one, let me know. Thanks!

> @stianst Please refrain from further processing this one for couple of mins yet, till I fixed the issue above. Thanks!

@stianst Should be OK to proceed this one further once @pedroigor confirms they are all set. Thanks!

> 
> > It is basically https://issues.redhat.com/browse/WFLY-14356, but applied to our revision caches and when setting up the embedded caches.
> > It should be enough to fix the unnecessary byte[] conversion issue that we found when running performance tests some time ago.
> > [1] https://github.com/pedroigor/keycloak/tree/KEYCLOAK-16724

**Comment by pedroigor** [comment]
@iankko LGTM now. Not sure about test failures though.

**Comment by pedroigor** [comment]
@tnorimat Thanks.

I forgot to mention the UI part. For other flows, including device flow, the settings for interval and token/code lifetimes are also set on a per-client basis. I think it makes sense to do that for CIBA too. I guess you might have different settings depending on the CD using Keycloak.

**Comment by pedroigor** [comment]
@tnorimat This is the branch actually, with my changes. https://github.com/pedroigor/keycloak/tree/KEYCLOAK-12137.

**Comment by iankko** [comment]
Superseded by https://github.com/keycloak/keycloak/pull/7857 one (all changes formerly present in this one were migrated to the successor).

**Pull request closed without merging**